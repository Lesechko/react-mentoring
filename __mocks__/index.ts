export * from './mockData'
export * from './mockRouter'
