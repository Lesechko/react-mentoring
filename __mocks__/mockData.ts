import { ISortingItem } from '../src/components/common/sorting/Sorting.constants'
import { IMovie } from '../src/components/movieList/movie/Movie'
import { IMovieState } from '../src/redux/reducers/moviesReducer'

export const movie: IMovie = {
  id: 238713,
  title: 'Spy',
  tagline: 'One of the guys. One of the Spies.',
  vote_average: 6.8,
  vote_count: 3058,
  release_date: '2015-05-06',
  poster_path:
    'https://image.tmdb.org/t/p/w500/xaOq0BnMHl445iZ4B7szBzaWgjS.jpg',
  overview:
    'A desk-bound CIA analyst volunteers to go',
  budget: 65000000,
  revenue: 235666219,
  genres: ['Action', 'Comedy', 'Crime'],
  runtime: 120,
}

export const newMovie = {
  id: 346,
  title: 'Seven Samurai',
  tagline:
    'The Mighty Warriors Who Became the Seven National Heroes of a Small Town',
  vote_average: 8.2,
  vote_count: 1088,
  release_date: '1954-04-26',
  poster_path:
    'https://image.tmdb.org/t/p/w500/v6xrz4fr92KY1oNC3HsEvrsvR1n.jpg',
  overview:
    "A samurai answers a village's request for protection after he falls on hard times. ",
  budget: 2000000,
  revenue: 271841,
  genres: ['Action', 'Drama'],
  runtime: 207,
}

export const mockMovieState: IMovieState = {
  movieList: [movie],
  activeMovie: movie,
}


export const sortingItem: ISortingItem[] = [
  { id: 1, value: 'release_date', title: 'release date' },
  { id: 2, value: 'runtime', title: 'runtime' },
  { id: 3, value: 'vote_average', title: 'rating' },
]
