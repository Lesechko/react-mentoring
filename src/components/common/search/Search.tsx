import { Button, BUTTON_TYPES } from '../button'
import { ReactElement, useState } from 'react'

import styles from './Search.css'

interface ISearch {
  onSearch(string: string): void
  placeholder: string
  actionName: string
}

const Search = ({
  onSearch,
  placeholder,
  actionName,
}: ISearch): ReactElement => {
  const [value, setValue] = useState('')

  return (
    <div className={styles.search}>
      <input
        placeholder={placeholder}
        className={styles.input}
        value={value}
        onChange={(e) => setValue(e.target.value)}
      />
      <Button type={BUTTON_TYPES.COMPLETE} onClick={() => onSearch(value)}>
        {actionName}
      </Button>
    </div>
  )
}

export default Search
