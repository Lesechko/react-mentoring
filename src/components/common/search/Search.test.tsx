import Search from './Search'
import { createMemoryHistory } from 'history'
import { fireEvent, render, screen } from '@testing-library/react'
import { withRouter } from '../../../../__mocks__'

const PLACEHOLDER = 'placeholder'
const ACTION_NAME = 'placeholder'
const VALUE = 'initial value'
let getByTestId: (name: string) => HTMLElement

beforeEach(() => {
  const Component = withRouter(
    <Search placeholder={PLACEHOLDER} actionName={ACTION_NAME} value={VALUE} />,
  )
  const component = render(Component)
  getByTestId = component.getByTestId
})

test('Should set value prop as input element valus', () => {
  const inputEl = getByTestId('input') as HTMLInputElement

  expect(inputEl.value).toBe(VALUE)
})

test('Should set placeholder prop as input element placeholder', () => {
  const inputEl = getByTestId('input') as HTMLInputElement

  expect(inputEl.placeholder).toBe(PLACEHOLDER)
})

test('Change value of input correctly', () => {
  const newValue = 'newMovie'
  const inputEl = getByTestId('input') as HTMLInputElement

  fireEvent.change(inputEl, { target: { value: newValue } })

  expect(inputEl.value).toEqual(newValue)
})

test('Search component match snapshot', () => {
  const Component = withRouter(
    <Search placeholder={PLACEHOLDER} actionName={ACTION_NAME} value={VALUE} />,
  )
  const { asFragment } = render(Component)

  expect(asFragment()).toMatchSnapshot()
})

test('Search component match snapshot', () => {
  const fn = jest.fn()
  const history = createMemoryHistory()
  const btn = screen.getByText(ACTION_NAME)
  fireEvent.click(btn, { target: { value: 'GGG' } })
  fn()
  expect(fn).toBeCalled()
})
