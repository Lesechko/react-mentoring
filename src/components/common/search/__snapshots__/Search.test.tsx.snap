// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component match snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <input
      class="input"
      data-testid="input"
      id="input"
      placeholder="placeholder"
      value="initial value"
    />
    <button
      class="btn btn-complete"
      data-testid="button"
      type="button"
    >
      placeholder
    </button>
  </div>
</DocumentFragment>
`;
